Src/Support/Google.Apis.Auth/OAuth2/TokenRefreshManager.cs

```diff
@@ -0,0 +1,170 @@
+                    return _token;
+                }
+            }
+            // The token may be set due to operations other than GetAccessTokenForRequestAsync, but we don't need to
+            // null out _refreshTask if so; it will be 
+            set
+            {
+                lock (_lock)
+                {
+                    _token = value;
+                }
+            }
+        }
+
+        internal async Task<string> GetAccessTokenForRequestAsync(string authUri, CancellationToken cancellationToken)
+        {
+            TokenResponse currentToken;
+            Task<TokenResponse> currentRefreshTask;
+            lock (_lock)
+            {
+                currentToken = _token;
+                currentRefreshTask = _refreshTask;
+            }
+            // Note: IsExpired is really "should be refreshed".
+            if (currentToken == null || currentToken.IsExpired(_clock))
+            {
+                // Start a refresh if nothing else is already refreshing the token.
+                // It's possible to end up with multiple requests here (if another thread started a task while
+                // we were checking for expiry) but that's a small window, and harmless so long as there aren't *too*
+                // many requests.
+                if (currentRefreshTask == null)
+                {
+                    currentRefreshTask = RefreshTokenAsync();
+                    lock (_lock)
+                    {
+                        _refreshTask = currentRefreshTask;
+                    }
+                }
+                // If the current token is still valid for a while, we can use it anyway. Otherwise,
+                // we need to wait for our refresh task to complete.
+                if (currentToken == null || currentToken.IsEffectivelyExpired(_clock))
+                {
+                    Task<TokenResponse> taskToAwait = currentRefreshTask;
+                    if (cancellationToken.CanBeCanceled)
+                    {
+                        // Reasonably simple way of creating a task that can be cancelled, based on another task.
+                        // (It would be nice if this were simpler.)
+                        taskToAwait = taskToAwait.ContinueWith(
+                            task => ResultWithUnwrappedExceptions(task),
+                            cancellationToken,
+                            TaskContinuationOptions.None,
+                            TaskScheduler.Default);
+                    }
+                    currentToken = await taskToAwait.ConfigureAwait(false);
+                }
+            }
+            return currentToken.AccessToken;
+        }
+
+
+        private async Task<TokenResponse> RefreshTokenAsync()
+        {
+            _logger.Debug("Token has expired, trying to get a new one.");
+            try
+            {
+                // TODO: Is using a cancellation task of "none" definitely appropriate? If this hangs forever,
```

Comment:
As suggested by @Ramkarthik in#1160, retry could be used here. In which case, the cancellation-token would be used to timeout each refresh request.